Feature/66343 comment fields for project attributes

toy opened the pull request:
# Ticket
https://community.openproject.org/wp/66343 (originally https://community.openproject.org/wp/64264)

# What are you trying to accomplish?
Allow to comment project attributes

# Leftovers

* Tests:
  * Feature tests for project overview (`modules/overviews/app/components/overviews/project_custom_fields/*` + `modules/overviews/app/controllers/overviews/project_custom_fields_controller.rb`)
  * `app/models/custom_field.rb` : test `can_have_comment?` and `comment_for`
  * `lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb` - tested implicitly, but better to test expected behaviour for project as having comment and work package as not having them
* API docs for the newly introduced custom comment fields
* Decide if custom comments functionality should stay in acts as customizable or be moved to `Projects::CustomFields`
* N+1 when exporting project query is still present

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

toy committed:
[#64264] Comment fields for project attributes

https://community.openproject.org/work_packages/64264

toy committed:
migration to add has_comment to custom_fields + custom comment model

toy committed:
manage has_comment

toy committed:
custom comment model

toy committed:
allow comments for all field formats

https://community.openproject.org/projects/stadt-koeln/work_packages/64264/activity#comment-1607511

toy committed:
add has_comment to field formats using primer form

toy committed:
access to CustomField#has_comment? from item component of project overview

Also use delegate for calculated_value?

toy committed:
handle has comment when choosing cf value wrapper (modal/comment/plain)

toy committed:
extract value_wrapper_attributes

toy committed:
handle showing readonly modal

toy committed:
handle displaying custom comment in project row component

toy committed:
CustomField#comment_for method

toy committed:
custom comment column in project list

toy committed:
Project#custom_comment_for method

toy committed:
exporting custom field comments

toy committed:
ensure custom comment has newlines normalised, to not deal with that in journal logic

toy committed:
custom comment journaling

toy committed:
make can_have_comment? a class method of CustomField and for instance method to rely on it

toy committed:
create custom comment journals only when custom field is enabled (for Project) and has has_comment mark

toy committed:
add custom comment changes

toy committed:
show custom comment journal diff

toy committed:
use same logic for hiding deleted custom field activity as for comments

Currently removing custom field will reveal its activity except for
project, for which current state of linking is used, which is deleted
alongside custom field itself

toy committed:
move admin_only check of custom field related activity to the query

This should be more efficient and will ensure "The changes were
retracted" message shown when all changes are hidden

toy committed:
allow showing journal diff for deleted custom fields

toy committed:
delegate CustomField#can_have_comment? instead of defining

toy committed:
adapt changes from ac1f346 and 4ab26f50 to custom comment journaling

toy committed:
add CustomField#comment_attribute_name

toy committed:
allow custom comment attributes through project update contract

toy committed:
refactor checking if field can be diffed

toy committed:
transform keys to strings instead of converting to hash with indifferent access

toy committed:
field for custom comment

toy committed:
extract DialogComponent for shared behaviour between Edit and Show dialogs

toy committed:
action and permission for dialog to show value'n'comment

toy committed:
permit custom comments for projects

toy committed:
inherit item component from show component allowing to render comment

toy committed:
move configuration for ability to have custom comments to acts_as_customizable

toy committed:
move adding custom_comments relation to acts_as_customizable and mark for autosave

toy committed:
allow to assign comments in a way similar to values

toy committed:
include acts as customizable methods at start of call

toy committed:
return changes to comments from custom_field_changes and changed_with_custom_fields

toy committed:
ensure custom comments are found if custom field is for all even wiithout custom_fields_projects rows

toy committed:
replace class needed only to set gap with gap-3

toy committed:
reduce gap between items in custom field overview sidebar

toy committed:
move custom_comment_for to acts as customizable

toy committed:
move contditions of calculated value errors property

toy committed:
comment value injector

toy committed:
comment_attribute_name should check if custom field has comment instead of if it can have comment

toy committed:
handle comment attribute in all contracts

toy committed:
handle writeableness of comments in project create contract

toy committed:
comment schema injector

toy committed:
replace availability joins with EXISTS condition for custom comment journaling

toy committed:
use non editable wrapper for all non editable fields

toy committed:
don't cache the scope in Project#all_available_custom_fields

toy committed:
rework projects index columns config helper

See 31ba520a70bef73df2a1632e1f190c3d9187ff51

toy committed:
test custom comment column in projects index

toy committed:
add class to custom comment cells like for description

In parallel checking if the class is needed at all

toy committed:
don't allow filtering by custom columns

toy committed:
test handling available custom fields in BaseContract

toy committed:
add accessors for custom comments

It was needed by project create contract spec

toy committed:
always return comment_attribute_name even when has_comment is false

toy committed:
add accessors for comments also when has_comment is false

toy committed:
set comments and show changes even if custom field doesn't allow comments

Otherwise contract doesn't know that there was an attempt to set
attribute

toy committed:
handle custom comment human attribute name

toy committed:
rename shared example in project create contract

toy committed:
small refactor of project attribute specs

toy committed:
rework setting custom comments to allow setting them also for inacessible custom fields

This should be blocked on contract level

toy committed:
test custom comments in project create contract

toy committed:
don't add nesting with shared examples when not needed in update contract spec

toy committed:
a bit of rubocop corrections

toy committed:
clearly distinguish shared spec for attributes and custom field values

toy committed:
test writable attributes for custom comments in update contract

toy committed:
cleanup permissions part of update contract spec

toy committed:
test permissions for custom comments in update contract

toy committed:
prevent revealing of deleted custom fields diff only for journable that can have admin_only fields

toy committed:
add admin_only_allowed setting to acts_as_customizable

toy committed:
shorten custom value assignment in journals controller spec

toy committed:
extract shared example for when diff is shown

toy committed:
cleanup testing project custom fields in journals controller

toy committed:
test showing journal diff for deleted custom fields

toy committed:
test showing diff of custom comments

toy committed:
test custom comment form field

toy committed:
test custom comment injection into api

toy committed:
use wait_for_network_idle instead of wait_for_reload in projects index spec

toy committed:
use complete label for comment field if there are multiple custom fields rendered

toy committed:
cleanup additional css classes for project index columns

Usage of project-long-text-container is nowhere to be found

toy committed:
test rendering custom comments fields

toy committed:
test custom value inputs using public method

toy committed:
test CustomField comment attribute name methods

toy committed:
use bang versions of update/save/destroy in project acts_as_journalized tests

toy committed:
auto enable custom field for project when creating custom comment

toy committed:
test journaling custom comments

toy committed:
test custom comments journal formatter

toy committed:
rework test for journaling custom values

toy committed:
test Redmine::Acts::Customizable::HumanAttributeName

toy committed:
remove auto enabling custom field for project when creating custom comment

dombesz commented on app/forms/custom_fields/custom_field_rendering.rb:
```diff
@@ -77,6 +82,15 @@ def custom_field_input(builder, custom_field)
+  def custom_comment_input(builder, custom_field)
+    CustomFields::CommentField.new(
+      builder,
+      custom_field: custom_field,
```
```suggestion
      custom_field:,
```

dombesz commented on app/forms/custom_fields/custom_field_rendering.rb:
```diff
@@ -77,6 +82,15 @@ def custom_field_input(builder, custom_field)
+      builder,
+      custom_field: custom_field,
+      object: model,
+      complete_label: custom_fields.length > 1
```
I'm wondering if we still have a case where multiple project attributes can be edited on the same form? We do have a few places, like creating a new projects, copying projects and creating from template, but none of those places currently display the comments anyway.

toy replied:
This was not specified, but I assumed that if custom value is editable, so should be the comment, so creating a project from template is where all custom fields with comments will show the comment field
<img width="794" height="893" alt="image" src="https://github.com/user-attachments/assets/fdc3f087-bf67-43ca-aece-4747f0807ca6" />

dombesz replied:
~~Thanks, unfortunately the custom comments are not being saved, the `params.new_project` needs to permit `custom_comments` too.~~
![Feb-24-2026 22-36-49](https://github.com/user-attachments/assets/e3db571d-fe0a-4baa-8bcf-a1fb2dac0eae)

Edit: I was about to report the issue above, but I see you just provided the fix for it. Thanks!

dombesz replied:
The comment fields seem to be cleared when there is a validation error, same happens when copying a project too. I'd say let's address this separately.

![Feb-24-2026 23-02-03](https://github.com/user-attachments/assets/49e4c367-5a8e-48ff-bfe4-6e9800b307ac)

toy replied:
So for custom comments saving on project creation - passing parameters (https://github.com/opf/openproject/commit/f7e001f6ea798ba934a5c41be8b5ce5dee0fa24e#diff-eb5213babf99c1c0cb3130ccaf08ab620683c4a07b261a00c9295f5287f4af4aL306-R307) was needed and for not loosing comment on failed validation - using association on object/project side instead of on custom field side (https://github.com/opf/openproject/pull/21452/changes/80f8785a06f2d22d6ce353748afb55dc893fd453)

toy replied:
Thank you for spotting

dombesz commented on app/helpers/projects_helper.rb:
```diff
@@ -49,7 +49,12 @@ def projects_sort_header_tag(column, all_column_attributes, **)
+      filter_column_mapping = filter_column_mapping.merge(column.attribute.to_s => nil)
+    end
+
+    sort_header_with_action_menu(column, all_column_attributes, filter_column_mapping, **,
```
The `Filter by` menu entry is still being shown, but it does not do anything.

<img width="257" height="280" alt="Image" src="https://github.com/user-attachments/assets/e4a1b2a7-13f9-48f2-9a84-ed222c4e4361" />

toy replied:
That is odd, the code in helper is exactly for that and it does work when I test it

dombesz replied:
Seems to be gone now, not sure what was causing it.

dombesz commented on app/models/exports/formatters/custom_comment.rb:
```diff
@@ -0,0 +1,53 @@
+
+      def find_custom_field(object)
+        id = attribute.to_s.delete_prefix("cfc_").to_i
+        object.available_custom_fields.find { it.id == id }
```
This call triggers an N+1 query, which results in a slow down from 8s to 30s when exporting ~2000 projects on my local machine. That is a 3.5x slow down.

```bash

 ↳ app/models/exports/formatters/custom_comment.rb:49:in 'Exports::Formatters::CustomComment#find_custom_field'
  CACHE ProjectCustomField Eager Load (0.0ms)  SELECT "custom_fields"."id" AS t0_r0, "custom_fields"."type" AS t0_r1, "custom_fields"."field_format" AS t0_r2, "custom_fields"."regexp" AS t0_r3, "custom_fields"."min_length" AS t0_r4, "custom_fields"."max_length" AS t0_r5, "custom_fields"."is_required" AS t0_r6, "custom_fields"."is_for_all" AS t0_r7, "custom_fields"."is_filter" AS t0_r8, "custom_fields"."position" AS t0_r9, "custom_fields"."searchable" AS t0_r10, "custom_fields"."editable" AS t0_r11, "custom_fields"."admin_only" AS t0_r12, "custom_fields"."multi_value" AS t0_r13, "custom_fields"."default_value" AS t0_r14, "custom_fields"."name" AS t0_r15, "custom_fields"."created_at" AS t0_r16, "custom_fields"."updated_at" AS t0_r17, "custom_fields"."content_right_to_left" AS t0_r18, "custom_fields"."allow_non_open_versions" AS t0_r19, "custom_fields"."custom_field_section_id" AS t0_r20, "custom_fields"."position_in_custom_field_section" AS t0_r21, "custom_fields"."formula" AS t0_r22, "custom_fields"."has_comment" AS t0_r23, "custom_field_sections"."id" AS t1_r0, "custom_field_sections"."position" AS t1_r1, "custom_field_sections"."name" AS t1_r2, "custom_field_sections"."type" AS t1_r3, "custom_field_sections"."created_at" AS t1_r4, "custom_field_sections"."updated_at" AS t1_r5, "custom_field_sections"."display_representation" AS t1_r6 FROM "custom_fields" LEFT OUTER JOIN "custom_field_sections" ON "custom_field_sections"."id" = "custom_fields"."custom_field_section_id" AND "custom_field_sections"."type" = $1 WHERE "custom_fields"."type" = $2 AND "custom_fields"."id" IN (SELECT "project_custom_field_project_mappings"."custom_field_id" FROM "project_custom_field_project_mappings" WHERE "project_custom_field_project_mappings"."project_id" = $3) ORDER BY custom_field_sections.position, "custom_fields"."position_in_custom_field_section" ASC  [["type", "ProjectCustomFieldSection"], ["type", "ProjectCustomField"], ["project_id", 2092]]
```

toy replied:
At least this method is identical for CustomField formatter, do you mean N+1 from custom comments?

dombesz replied:
Yes, the custom comments triggering the N+1 queries in the **background job**. Forgot to mention that, sorry!

toy replied:
https://github.com/opf/openproject/pull/21452/changes/e2ac1614c3b70962f4b9cace84f2d4aa8b1d2155 should take care of that, can you please check in your example?

dombesz replied:
Now I see that the comments are eager loaded in the background job, however the N+1 queries coming a bit later in the code. My setup exports multiple custom fields too, for 2214 projects and 7 comment fields enabled, I get 35463 queries.

<details><summary> 👉 Click to see relevant call stack.</summary>
<p>

```
app/models/exports/formatters/custom_comment.rb:50:in 'Exports::Formatters::CustomComment#find_custom_field'
app/models/exports/formatters/custom_comment.rb:41:in 'Exports::Formatters::CustomComment#retrieve_value'
app/models/exports/formatters/default.rb:53:in 'Exports::Formatters::Default#format'
app/models/exports/exporter.rb:77:in 'Exports::Exporter#format_attribute'
app/models/project/pdf_export/common/project_attributes.rb:168:in 'Project::PDFExport::Common::ProjectAttributes#process_attribute_field'
app/models/project/pdf_export/common/project_attributes.rb:88:in 'Project::PDFExport::Common::ProjectAttributes#process_field'
app/models/project/pdf_export/common/project_attributes.rb:68:in 'block in Project::PDFExport::Common::ProjectAttributes#collect_field_groups'
app/models/project/pdf_export/common/project_attributes.rb:67:in 'Array#each'
app/models/project/pdf_export/common/project_attributes.rb:67:in 'Enumerable#inject'
app/models/project/pdf_export/common/project_attributes.rb:67:in 'Project::PDFExport::Common::ProjectAttributes#collect_field_groups'
app/models/project/pdf_export/common/project_attributes.rb:58:in 'Project::PDFExport::Common::ProjectAttributes#write_project_detail_content_tables'
app/models/project/pdf_export/common/project_attributes.rb:38:in 'Project::PDFExport::Common::ProjectAttributes#write_project_detail_content'
app/models/projects/exports/pdf_export/report.rb:44:in 'block in Projects::Exports::PDFExport::Report#write_project_detail'
app/models/exports/pdf/common/common.rb:86:in 'block in Exports::PDF::Common::Common#with_margin'
app/models/exports/pdf/common/common.rb:92:in 'Exports::PDF::Common::Common#with_vertical_margin'
app/models/exports/pdf/common/common.rb:85:in 'Exports::PDF::Common::Common#with_margin'
app/models/projects/exports/pdf_export/report.rb:42:in 'Projects::Exports::PDFExport::Report#write_project_detail'
app/models/projects/exports/pdf_export/report.rb:36:in 'block in Projects::Exports::PDFExport::Report#render_report'
app/models/projects/exports/pdf_export/report.rb:34:in 'Array#each'
app/models/projects/exports/pdf_export/report.rb:34:in 'Projects::Exports::PDFExport::Report#render_report'
app/models/projects/exports/pdf.rb:146:in 'Projects::Exports::PDF#render_projects_report'
app/models/projects/exports/pdf.rb:124:in 'Projects::Exports::PDF#render_projects_report_pdf'
app/models/projects/exports/pdf.rb:97:in 'Projects::Exports::PDF#render_pdf'
app/models/projects/exports/pdf.rb:60:in 'Projects::Exports::PDF#export!'
app/workers/exports/export_job.rb:77:in 'Exports::ExportJob#export!'
app/workers/exports/export_job.rb:46:in 'block in Exports::ExportJob#perform'
lib/open_project/locale_helper.rb:36:in 'OpenProject::LocaleHelper.with_locale_for'
app/models/user.rb:546:in 'User.execute_as'
app/workers/exports/export_job.rb:44:in 'Exports::ExportJob#perform'
app/workers/shared_job_setup.rb:59:in 'block in SharedJobSetup#prepare_job_context'
app/workers/shared_job_setup.rb:76:in 'SharedJobSetup#with_clean_request_store'
app/workers/shared_job_setup.rb:55:in 'SharedJobSetup#prepare_job_context'
```

</p>
</details> 


![Feb-25-2026 18-00-19](https://github.com/user-attachments/assets/bf583aa5-69a8-4afe-b974-d82d43e9a74b)

toy committed:
integration tests for custom comments export

toy committed:
unit test for custom comment export formatter

dombesz commented on app/models/projects/custom_fields.rb:
```diff
@@ -60,7 +60,7 @@ def available_custom_fields
     # the extension to be completely removed from the acts_as_customizable plugin.
     def all_available_custom_fields
-      @all_available_custom_fields ||= ProjectCustomField
+      ProjectCustomField
```
I wonder if this was causing any issues? Although seems like, removing the variable cache is fine, because the value usually cached in the caller methods.

toy replied:
No, just bothering me, as it was caching only a simple scope

dombesz commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -126,7 +136,7 @@ def custom_fields=(values)
           return unless values.is_a?(Hash) && values.any?
 
-          values.with_indifferent_access.each do |custom_field_id, new_values|
+          values.transform_keys(&:to_s).each do |custom_field_id, new_values|
```
```suggestion
          values.stringify_keys.each do |custom_field_id, new_values|
```

toy replied:
Curiosity led me to learn why `stringify_keys` is just a bit more than `transform_keys(&:to_s)`: https://github.com/rails/rails/commit/8c7e69b79b63a88a170a9b9004a906db00161a3b

dombesz replied:
Not sure what happened here, are we not using `stringify_keys` here anymore?

toy replied:
Good catch, forgot to pull the change made on github before force pushing 🤦

dombesz commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -483,6 +537,20 @@ def remove_custom_value(custom_value)
 
+        def set_custom_comments(values:, comments_by_field_id:)
+          values.each do |custom_field_id, text|
+            custom_field_id = custom_field_id.to_s.to_i
```
I'm wondering if we need `to_s` here?

toy replied:
Sadly yes, as `custom_field_id` can be a Symbol, I probably better add a comment to clarify it

dombesz commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -387,21 +436,18 @@ def build_custom_value(custom_field, value:)
-              yield custom_field
-            end
-          end
+          return unless /\Acustom_(?:field|comment)_(?<id>\d+)=?\z/ =~ method_symbol.to_s
```
💡TIL: the named capture group `<id>` will be assigned as a local variable `id`.

I'm not sure how to feel about its discoverability tho, especially that it shadows the `id` method on the object.

toy replied:
I'm also not sure if it is more expressive or more hard to read, I'll switch to using `.match` instead of `=~`

dombesz commented on lib_static/plugins/acts_as_customizable/lib/human_attribute_name.rb:
```diff
@@ -37,10 +37,14 @@ module HumanAttributeName
         def human_attribute_name(attribute, options = {})
-          match = /\Acustom_field_(?<id>\d+)\z/.match(attribute)
+          /\Acustom_(?<type>field|comment)_(?<id>\d+)\z/ =~ attribute
+          name = CustomField.find_by(id:).name if id
```
The implicit variable assignment is definitely harder to understand here.

toy replied:
I'll switch to using `.match`

dombesz reviewed (commented):
Works as expected, great job overall! I left a few notes, but none of them really critical. I am still skimming through the specs, just making my comments visible.

dombesz commented on lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -52,6 +54,13 @@ def acts_as_customizable(options = {})
              validate: false,
              autosave: true
 
+          if can_have_custom_comments?
```
If we don't want to make comments available for every custom field, then perhaps would be better to move the custom comments related extensions to the [Projects::CustomFields](https://github.com/opf/openproject/blob/dev/app/models/projects/custom_fields.rb) concern.

toy replied:
I'll put it up for discussion after merging

toy committed:
test presence of UTF-8 BOM mark in CSV separately

toy committed:
reuse defined permissions when testing without view_project_attributes

dombesz commented on spec/contracts/projects/create_contract_spec.rb:
```diff
@@ -222,12 +222,32 @@
 
-        shared_examples "can read project attributes" do
-          it "can read project attributes" do
+        shared_examples "can access custom field" do
```
Shall we have a `"can access custom field comment"` shared example?

toy replied:
It is about accessing custom field, so both its value and comment, so not really. Unfortunately custom field very often means custom value, so it is not always clear that here it really means custom field itself

dombesz commented on spec/contracts/projects/update_contract_spec.rb:
```diff
@@ -34,119 +34,177 @@
     end
 
     describe "permissions" do
+      name_parent_identifier_readonly = %i[name parent_id identifier].index_with { %i[error_readonly] }
```
Just a nitpick, but this variable name does not inform me that it is an error hash. How about `readonly_attribute_errors`? Having the hash explicitly typed out also makes it easier to read.

```suggestion
      let(:readonly_attribute_errors) do
          { name: %i(error_readonly),
            parent_id: %i(error_readonly),
            identifier: %i(error_readonly) }
      end
```

toy replied:
I tried to compact the repetition in error, but also felt doubt that it is clearly readable
I also reworked the specs to allow `let` to be used, as variable still looks odd in specs
https://github.com/opf/openproject/pull/21452/changes/46bf7ab8e63980ae99ac4213efa31ebd99c16a6b

toy committed:
make project created for version as inactive to remove it from query

toy committed:
test pdf export with custom fields by admin

dombesz commented on spec/lib/open_project/journal_formatter/custom_comment_spec.rb:
```diff
@@ -0,0 +1,118 @@
+  let(:full_link) { link_to(I18n.t(:label_details), url, class: "diff-details", target: "_top") }
+
+  let(:expected) do
+    I18n.t(i18n_key, label:, link:)
```
Just a  nitpick, but how about this:

```suggestion
    I18n.t(i18n_key, label: expected_label, link: expected_link)
```

toy committed:
don't use local variables automatically populated for named capture groups

toy committed:
handle delete custom fields in human attribute name

dombesz commented on spec/models/projects/exporter/csv_integration_spec.rb:
```diff
@@ -127,14 +128,14 @@
 
       it "renders all globally available project custom fields including hidden ones in the header" do
-        expect(parsed.size).to eq 3
+        expect(parsed.size).to eq 2
```
I'm wondering why is the number of returned projects is returned ?

toy replied:
There was an extra project created for version and I wanted to get rid of it primarily to not complicate the spec for pdf export by an admin (https://github.com/opf/openproject/pull/21452/changes/0dd09623d70821053bec6d363da73d0bf534675a#diff-13ae470b1bd6bc9179f75d0eee6f9fc03cd1cae9e2232d9db8a3d847453f6f0f), the simplest way was to archive it, but then it is not present in other exports too, leading to this change, see
https://github.com/opf/openproject/pull/21452/changes/acb67ca9a398590d17c4cdc5418338edd0065a2a#diff-1a04518b413afcacdfd393517fceff168bf1d096a446135ae8a294e4423575a3

dombesz commented on spec/models/projects/exporter/csv_integration_spec.rb:
```diff
@@ -158,6 +159,70 @@
+    context "without view_project_attributes permission" do
+      let(:permissions) { super() - %i[view_project_attributes] }
+
+      it "does not render project custom fields in the header" do
```
The title is a bit misleading because it doesn't mention the comments are not being returned either.

toy replied:
Nice catch, updating some descriptions slipped my attention

dombesz commented on spec/models/projects/exports/pdf_spec.rb:
```diff
@@ -149,14 +153,92 @@ def get_test_column_value(custom_field)
+          project.name,
+          "ID", project.id.to_s,
+          "#{version_cf.name} comment", "Comment visible to members",
+          "#{hidden_cf.name} comment", "Comment visible to admins",
```
Shall we have a comment field for the `not_used_string_cf` to ensure that it's not accidentally included?

toy replied:
It is already tested as whole document is compared, but I can add it explicitly

dombesz reviewed (approved):
Went through the specs in the second round of the review, left a few minor notes/questions. One thing that I believe is missing is the API docs for the newly introduced custom comment fields. I'm fine with addressing that in a separate PR.

LGTM!✅

toy committed:
use the object.custom_comment_for(cf) instead of cf.comment_for(object)

As comments are built on object size, they will be present after failed
validation, but for comments from the custom field side they will be
queried from DB, so will not be the built ones

toy committed:
eager load custom comments in query exporter

toy committed:
rework readonly_attribute_errors variable in project update contract spec

dombesz reviewed (commented):
The create project form works as expected, but unfortunately the N+1 query still persists. See my comment on the [relevant thread](https://github.com/opf/openproject/pull/21452#discussion_r2853265910).

On a side note, the existing comments from PIR templates are not carried over to the created project. (out of scope for this PR)
<details> <summary>Screenshots from template project</summary>
<p>
<img width="728" height="661" alt="Image" src="https://github.com/user-attachments/assets/42f8d18f-95c3-4223-a119-ab364c9d1a48" />
</p>
</details>

<details> <summary>Screenshots from the created project</summary>
<p>

<img width="1203" height="775" alt="Image" src="https://github.com/user-attachments/assets/01c42079-d1aa-4751-8624-772c6ea98a64" />
</p>
</details>

toy committed:
Use stringify_keys instead of transform_keys(&:to_s)

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

toy commented on lib/api/v3/utilities/custom_field_injector.rb:
```diff
@@ -276,10 +289,23 @@ def inject_property_value(custom_field, config)
+        end
+
+        def inject_comment_value(custom_field, config)
+          return unless custom_field.has_comment?
+
+          @class.property custom_field.comment_attribute_name.to_sym,
+                          if: ->(*) { available_custom_fields.include?(custom_field) },
+                          getter: ->(*) { custom_comment_for(custom_field)&.text },
+                          setter: ->(fragment:, **) { self.custom_comments = { custom_field.id => fragment } },
                           cache_if: config[:cache_if],
-                          render_nil: false
+                          render_nil: true
```
No then conversion happens automatically

toy replied:
This is done automatically

toy committed:
consistent camel casting custom field injector fields, otherwise it is done automatically

toy commented on app/models/custom_field.rb:
```diff
@@ -369,6 +390,11 @@ def first_calculation_error(customized)
   end
 
+  def comment_for(customized)
+    # Use a ruby finder following same logic as in first_calculation_error
+    comments.find { it.customized == customized }
```
Two separate methods to be able to clearly decide on which side to load comments if needed eagerly

toy commented:
@dombesz Thank you for review, I'll create tickets or notes to discuss for the leftovers
